pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeFieldStep.java

```diff
@@ -0,0 +1,153 @@
+    if (context.getValueSchema().getSchemaInfo().getType() == AVRO) {
+      GenericRecord avroRecord = (GenericRecord) context.getValueObject();
+      GenericRecord newRecord = computeFields(fields, avroRecord, valueSchemaCache, context);
+      if (avroRecord != newRecord) {
+        context.setValueModified(true);
+      }
+      context.setValueObject(newRecord);
+    }
+  }
+
+  public void computeKeyFields(List<ComputeField> fields, TransformContext context) {
+    if (context.getKeyObject() != null
+        && context.getValueSchema().getSchemaInfo().getType() == AVRO) {
+      GenericRecord avroRecord = (GenericRecord) context.getKeyObject();
+      GenericRecord newRecord = computeFields(fields, avroRecord, keySchemaCache, context);
+      if (avroRecord != newRecord) {
+        context.setKeyModified(true);
+      }
+      context.setKeyObject(newRecord);
+    }
+  }
+
+  private GenericRecord computeFields(
+      List<ComputeField> fields,
+      GenericRecord record,
+      Map<org.apache.avro.Schema, org.apache.avro.Schema> schemaCache,
+      TransformContext context) {
+    org.apache.avro.Schema avroSchema = record.getSchema();
+
+    List<Schema.Field> computedFields =
+        fields
+            .stream()
+            .map(f -> new org.apache.avro.Schema.Field(f.getName(), getAvroSchema(f.getType())))
+            .collect(Collectors.toList());
+
+    Set<String> computedFieldNames =
+        computedFields.stream().map(f -> f.name()).collect(Collectors.toSet());
+    // New fields are the intersection between existing fields and computed fields. Computed fields
+    // take precedence.
+    List<Schema.Field> newFields =
+        avroSchema
+            .getFields()
+            .stream()
+            .filter(f -> !computedFieldNames.contains(f.name()))
+            .map(
+                f ->
+                    new org.apache.avro.Schema.Field(
+                        f.name(), f.schema(), f.doc(), f.defaultVal(), f.order()))
+            .collect(Collectors.toList());
+    newFields.addAll(computedFields);
+    org.apache.avro.Schema newSchema =
+        schemaCache.computeIfAbsent(
+            avroSchema,
+            schema ->
+                org.apache.avro.Schema.createRecord(
+                    avroSchema.getName(),
+                    avroSchema.getDoc(),
+                    avroSchema.getNamespace(),
+                    avroSchema.isError(),
+                    newFields));
+
+    GenericRecord newRecord = new GenericData.Record(newSchema);
+    // Add original fields
+    for (org.apache.avro.Schema.Field field : avroSchema.getFields()) {
+      newRecord.put(field.name(), record.get(field.name()));
+    }
+    // Add computed fields
+    for (ComputeField field : fields) {
+      newRecord.put(field.getName(), field.getEvaluator().evaluate(context));
+    }
+    return newRecord;
+  }
+
+  private Schema getAvroSchema(ComputeFieldType type) {
+    switch (type) {
```

Comment:
Seems that you can map the ComputeFieldType to Schema.Type and do the `fieldTypeToAvroSchemaCache.computeIfAbsent(
            type, (ignored) -> Schema.create(Schema.Type.STRING))` after the `switch`